runtime/connectors/blob/blobdownloader.go

```diff
@@ -1,94 +1,385 @@
 )
 
+// increasing this limit can increase speed ingestion
+// but may increase bottleneck at duckdb or network/db IO
+// set without any benchamarks
+const concurrentBlobDownloadLimit = 8
+
+// map of supoprted extensions for partial downloads vs readers
+var partialDownloadReaders = map[string]string{".parquet": "parquet", ".csv": "csv", ".tsv": "csv", ".txt": "csv", ".parquet.gz": "parquet"}
+
+type Strategy string
+
+const (
+	TAIL Strategy = "tail"
+	HEAD Strategy = "head"
+	NONE Strategy = "none"
+)
+
+type ExtractConfig struct {
+	Strategy Strategy
+	Size     int64
+}
+
+type ExtractPolicy struct {
+	Row  ExtractConfig
+	File ExtractConfig
+}
+
+type BlobIterator struct {
+	bucket  *blob.Bucket
+	objects []*blobObject
+	index   int
+}
+
+func NewExtractConfigs(input *runtimev1.Source_ExtractPolicy) (*ExtractPolicy, error) {
+	config := &ExtractPolicy{
+		Row:  ExtractConfig{Strategy: NONE},
+		File: ExtractConfig{Strategy: NONE},
+	}
+
+	if input == nil {
+		return config, nil
+	}
+
+	// parse file
+	if input.File != nil {
+		// parse strategy
+		strategy, err := parseStrategy(input.File.Strategy)
+		if err != nil {
+			return nil, err
+		}
+
+		config.File.Strategy = strategy
+
+		// parse size
+		size, err := strconv.ParseInt(input.File.Size, 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("invalid size, parse failed with error %w", err)
+		}
+		if size <= 0 {
+			return nil, fmt.Errorf("invalid size %q", size)
+		}
+
+		config.File.Size = size
+	}
+
+	// parse rows
+	if input.Row != nil {
+		// parse strategy
+		strategy, err := parseStrategy(input.Row.Strategy)
+		if err != nil {
+			return nil, err
+		}
+
+		config.Row.Strategy = strategy
+
+		// parse size
+		// todo :: add support for number of rows
+		size, err := getBytes(input.Row.Size)
+		if err != nil {
+			return nil, fmt.Errorf("invalid size, parse failed with error %w", err)
+		}
+		if size <= 0 {
+			return nil, fmt.Errorf("invalid size %q", size)
+		}
+
+		config.Row.Size = size
+	}
+
+	return config, nil
+}
+
+func parseStrategy(s string) (Strategy, error) {
+	switch strings.ToLower(s) {
+	case "tail":
+		return TAIL, nil
+	case "head":
+		return HEAD, nil
+	default:
+		return "", fmt.Errorf("invalid extract strategy %q", s)
+	}
+}
+
+func getBytes(size string) (int64, error) {
+	var s datasize.ByteSize
+	if err := s.UnmarshalText([]byte(size)); err != nil {
+		return 0, err
+	}
+
+	return int64(s.Bytes()), nil
+}
+
 type FetchConfigs struct {
 	GlobMaxTotalSize      int64
 	GlobMaxObjectsMatched int
 	GlobMaxObjectsListed  int64
 	GlobPageSize          int
+	Extract               *ExtractPolicy
 }
 
-// downloads file to local paths
-// todo :: return blob handler as iterator
-func FetchFileNames(ctx context.Context, bucket *blob.Bucket, config FetchConfigs, globPattern, bucketPath string) ([]string, error) {
-	validateConfigs(&config)
-	prefix, glob := doublestar.SplitPattern(globPattern)
+// container encapsulates all extract policy
+type container struct {
```

Comment:
Container is called container because it **keeps items** as per extract policy. I will add better doc to justify the naming here.

I do agree with your point that container is little complicated at this point. 
But a lot of complexity is because of the fact that implementation of row strategy is dependent on whether partition strategy is present or not. 

I can separate the headPlanner and tailPlanner for file strategy (in fact that's what I had initially done ) but that doesn't reduce the complexity much because of the point mentioned above. I will anyways also share that solution with you.

The list.List is being used here since I don't want to keep all the objects in memory for tail listing(since we are always interested in the last n objects only). May be the usage will become more clear once I separate those planners.
An iterator instead of returning the entire list seems overkill at this point. We can move to that solution if we decide to implement tail strategy in depth first listings may be.